
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.3.5">
    
    
      
        <title>Commandline Interface - ADToolbox</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4a0965b7.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#adtoolbox-commandline-interface" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ADToolbox" class="md-header__button md-logo" aria-label="ADToolbox" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ADToolbox
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Commandline Interface
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ADToolbox" class="md-nav__button md-logo" aria-label="ADToolbox" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    ADToolbox
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Welcome to ADToolbox
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../Installation/" class="md-nav__link">
        Installation
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Commandline Interface
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Commandline Interface
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#initialization" class="md-nav__link">
    Initialization
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#adtoolbox-modules" class="md-nav__link">
    ADToolbox Modules
  </a>
  
    <nav class="md-nav" aria-label="ADToolbox Modules">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-configs-module" class="md-nav__link">
    1. Configs Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-database-module" class="md-nav__link">
    2. Database Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-metagenomics-module" class="md-nav__link">
    3. Metagenomics Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-adm-module" class="md-nav__link">
    4. ADM Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-documentations-module" class="md-nav__link">
    5. Documentations Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-report-module" class="md-nav__link">
    6. Report Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-utility-module" class="md-nav__link">
    7. Utility Module
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../API/" class="md-nav__link">
        API
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../Parameter_Tuning/" class="md-nav__link">
        Parameter Tuning
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#initialization" class="md-nav__link">
    Initialization
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#adtoolbox-modules" class="md-nav__link">
    ADToolbox Modules
  </a>
  
    <nav class="md-nav" aria-label="ADToolbox Modules">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-configs-module" class="md-nav__link">
    1. Configs Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-database-module" class="md-nav__link">
    2. Database Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-metagenomics-module" class="md-nav__link">
    3. Metagenomics Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-adm-module" class="md-nav__link">
    4. ADM Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-documentations-module" class="md-nav__link">
    5. Documentations Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-report-module" class="md-nav__link">
    6. Report Module
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-utility-module" class="md-nav__link">
    7. Utility Module
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="adtoolbox-commandline-interface">ADToolbox Commandline Interface</h1>
<p>Here we go over using the commandline interface, CLI, of ADToolbox.
First, we need to initialize the CLI:</p>
<h2 id="initialization">Initialization</h2>
<p>After installing ADToolbox, type and execute the following in your terminal to initialize the base directory for ADToolbox files:</p>
<pre><code>ADToolbox

</code></pre>
<p>You should see the following if you are running this for the first time:</p>
<pre><code>No Base Directory Found: 
Where do you want to store your ADToolBox Data?:

</code></pre>
<p>Type the <strong>absolute path</strong> directory of interest. Don't worry if you mess this part up. You can change this later as well. you can type '.' for now and change this later.</p>
<p>You can access all the commands along with their brief explanation by:</p>
<pre><code>ADToolbox --help

</code></pre>
<h2 id="adtoolbox-modules">ADToolbox Modules</h2>
<p>This toolbox is comprised of different modules:</p>
<ol>
<li>
<p>Configs Module</p>
</li>
<li>
<p>Database Module</p>
</li>
<li>
<p>Metagenomics Module</p>
</li>
<li>
<p>ADM Module</p>
</li>
<li>
<p>Documentations Module</p>
</li>
<li>
<p>Report Module</p>
</li>
<li>
<p>Utility Module</p>
</li>
</ol>
<hr />
<h3 id="1-configs-module">1. Configs Module</h3>
<p>After installation, you have to download all the required files to run ADToolbox properly. To do this, first go to the Configs module by:</p>
<pre><code>
ADToolbox Configs --help

</code></pre>
<p>This will give you a list of all functionalities that are related to the configurations of the toolbox. Here we go one by one in the correct order:</p>
<ul>
<li>set-base-dir:  The first configuration command will allow you to set the base directory for ADToolbox to work. This could be an existing folder somewhere in your files or a directory that you are willing to create. If the directory does not already exit, it will be automatically created after this command. For example if I want to set the base directory to be ADToolbox directory on my desktop the command would be, in MacOS, something like this:</li>
</ul>
<pre><code>
ADToolbox Configs set-base-dir -d ~/Desktop/ADToolbox

</code></pre>
<p>Note that you <strong><em>must</em></strong> include -d after <code>set-base-dir</code>.</p>
<p>Anything that you will do from now on, will be saved in this directory.</p>
<ul>
<li>build-folder-structure: Now you need to build the folder structure that is understandable by ADToolbox. You will do that by:</li>
</ul>
<pre><code>
ADToolbox Configs build-folder-structure

</code></pre>
<ul>
<li>download-all-databases: The next step is to download all the necessary database files for ADToolbox to work properly. You will achieve this by: </li>
</ul>
<pre><code>ADToolbox Configs download-all-databases

</code></pre>
<ul>
<li>download-escher-files: If you want to access the esher map functionalities of ADToolbox you need to download the required files by:</li>
</ul>
<pre><code>
ADToolbox Configs download-escher-files

</code></pre>
<p>An overal view of the Configs module of ADToolbox can be obtained by:</p>
<pre><code>
ADToolbox Configs --help

────────────────────────────────── ADToolBox───────────────────────────────────
usage: ADToolBox Configs [-h]
                         {set-base-dir,build-folder-structure,download-all-datab
ases,download-escher-files}
                         ...

positional arguments:
  {set-base-dir,build-folder-structure,download-all-databases,download-escher-fi
les}
                        Available Configs Commands:
    set-base-dir        Determine the address of the base directory for
                        ADToolBox to work with
    build-folder-structure
                        Builds the folder structure for ADToolBox to work
                        properly
    download-all-databases
                        Downloads all the databases for ADToolBox to work
                        properly, and puts them in the right directory in
                        Databases
    download-escher-files
                        Downloads all files required for running the escher
                        map functionality of ADToolBox

options:
  -h, --help            show this help message and exit


</code></pre>
<h3 id="2-database-module">2. Database Module</h3>
<p>Any database that is used by ADToolbox can be modified from this module. Type the following in your commandline to find all of the database module's commands:</p>
<pre><code>ADToolbox Database --help
──────────────────────────── ADToolBox ────────────────────────────
usage: ADToolBox Database [-h]
                          {initialize-feed-db,extend-feed-db,show-f
eed-db,download-reaction-db,download-seed-reaction-db,build-protein
-db,download-feed-db,download-protein-db,download-amplicon-to-genom
e-dbs}
                          ...

positional arguments:
  {initialize-feed-db,extend-feed-db,show-feed-db,download-reaction
-db,download-seed-reaction-db,build-protein-db,download-feed-db,dow
nload-protein-db,download-amplicon-to-genome-dbs}
                        Database Modules:
    initialize-feed-db  Initialize the Feed DB
    extend-feed-db      Extend the Feed Database using a CSV file
    show-feed-db        Display the Current Feed Database
    download-reaction-db
                        Downloads the reaction database in CSV
                        format
    download-seed-reaction-db
                        Downloads the seed reaction database in
                        JSON format
    build-protein-db    Generates the protein database for
                        ADToolbox
    download-feed-db    Downloads the feed database in JSON
                        format
    download-protein-db
                        Downloads the protein database in fasta
                        format; You can alternatively build it
                        from reaction database.
    download-amplicon-to-genome-dbs
                        downloads amplicon to genome databases

options:
  -h, --help            show this help message and exit
</code></pre>
<p>We now go over these commands one by one:</p>
<ul>
<li>initialize-feed-db: This will create an empty JSON file in the Database sub-directory in your base directory that will hold all the future feed information that you add. You can run this command  by:</li>
</ul>
<pre><code>
ADToolbox Database initialize-feed-db

</code></pre>
<ul>
<li>extend-feed-db: you can add to your current feed database from a csv file with this command. The CSV file <strong>must</strong> follow this column configuration:</li>
</ul>
<table>
<thead>
<tr>
<th>Name</th>
<th>TS</th>
<th>TSS</th>
<th>Lipids</th>
<th>Proteins</th>
<th>Carbohydrates</th>
<th>PI</th>
<th>SI</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>You can add your CSV file by:</p>
<pre><code>
ADToolbox Database extend-feed-db -d [PATH TO YOUR CSV FILE]

</code></pre>
<ul>
<li>show-feed-db: You can pretty print your current feed table by</li>
</ul>
<pre><code>
ADToolbox Database show-feed-db

</code></pre>
<p>This command will print a pretty table of your current feed database, which is the JSON file you initialized earlier and extended with a CSV file. This is made possible because of the great Rich library.</p>
<p><em>NOTE</em>: Skipp the following download commands if you ran <code>ADToolbox Configs download-all-databases</code>
- download-reaction-db: As the name implies, this will download the ADToolbox reaction database. This is required for many important modules of the toolbox</p>
<pre><code>ADToolbox Database download-reaction-db

</code></pre>
<ul>
<li>download-feed-db: Downloads the default feed database in JSON format</li>
</ul>
<pre><code>
ADToolbox Database download-feed-db

</code></pre>
<ul>
<li>download-protein-db: Downloads the protein database in fasta format; You can alternatively build it from reaction database if you have downloaded it; Check below.</li>
</ul>
<pre><code>
ADToolbox Database download-protein-db

</code></pre>
<ul>
<li>build-protein-db: Generates the protein database for ADToolbox from the reaction database:</li>
</ul>
<pre><code>
ADToolbox Database build-protein-db

</code></pre>
<ul>
<li>download-amplicon-to-genome-dbs: If you need to use the 16s mapping to the protein database and ADM, you will need to download the required databases using this command:</li>
</ul>
<pre><code>
ADToolbox Database download-amplicon-to-genome-dbs

</code></pre>
<ul>
<li>download-seed-reaction-db: This will download the SEED reaction database in JSON format.</li>
</ul>
<pre><code>ADToolbox Database download-seed-reaction-db

</code></pre>
<h3 id="3-metagenomics-module">3. Metagenomics Module</h3>
<p>Metagenomics module of ADToolbox is designed to add metagenomics data into considewration when designing and AD process.</p>
<p>You can observe all the functionalities by:</p>
<pre><code>
ADToolbox Metagenomics --help   

──────────────────────────── ADToolBox ────────────────────────────
usage: ADToolBox Metagenomics [-h]
                              {amplicon-to-genome,align-genomes,mak
e-json-from-genomes,map-genomes-to-adm}
                              ...

positional arguments:
  {amplicon-to-genome,align-genomes,make-json-from-genomes,map-geno
mes-to-adm}
                         Available Metagenomics Commands:
    amplicon-to-genome  Downloads the representative genome from
                        each amplicon
    align-genomes       Align Genomes to the protein database of
                        ADToolbox, or any other fasta with
                        protein sequences
    make-json-from-genomes
                        Generates JSON file required by Align-
                        Genomes for custom genomes.
    map-genomes-to-adm  maps JSON file with genome infromation to
                        ADM reactions

options:
  -h, --help            show this help message and exit

</code></pre>
<p>As of right now, there are 3 main submodules exist in the Metagenomics module:</p>
<ul>
<li>amplicon-to-genome: If you have 16s Data from QIIME, you can, hopefully, find the representative genomes for each replicon in an automated way
using this functionality by selecting different parameters:</li>
</ul>
<pre><code>ADToolbox Metagenomics amplicon-to-genome --help

────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox Metagenomics amplicon-to-genome [-h] [-q QIIME_OUTPUTS_DIR]
                                                 [-f FEATURE_TABLE_DIR]
                                                 [-r REP_SEQ_DIR]
                                                 [-t TAXONOMY_TABLE_DIR]
                                                 [-o OUTPUT_DIR]
                                                 [-a AMPLICON_TO_GENOME_DB]
                                                 [--k K]
                                                 [--similarity SIMILARITY]

options:
  -h, --help            show this help message and exit
  -q QIIME_OUTPUTS_DIR, --qiime-outputs-dir QIIME_OUTPUTS_DIR
                        Input the directory to the QIIME outputs
  -f FEATURE_TABLE_DIR, --feature-table-dir FEATURE_TABLE_DIR
                        Input the directory to the feature table output from
                        QIIME output tables
  -r REP_SEQ_DIR, --rep-Seq-dir REP_SEQ_DIR
                        Input the directory to the repseq fasta output from
                        QIIME output files
  -t TAXONOMY_TABLE_DIR, --taxonomy-table-dir TAXONOMY_TABLE_DIR
                        Input the directory to the taxonomy table output from
                        QIIME output files
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        Output the directory to store the representative
                        genome files
  -a AMPLICON_TO_GENOME_DB, --amplicon-to-genome-db AMPLICON_TO_GENOME_DB
                        The Amplicon to Genome Database to use
  --k K                 Top k genomes to be selected
  --similarity SIMILARITY
                        Similarity cutoff in the 16s V4 region for genome
                        selection

</code></pre>
<p>If you do not provide any arguents, the default directories and values will be used. By default, ADToolbox looks at
your base directory, that you set in Configs, in :<code>Metagenomics_Data/QIIME_Outputs</code>. </p>
<ul>
<li>
<p>If you want to use your default arguments, you need to provide the directory to Feature table, taxonomy table, repseq fasta file from QIIME.</p>
</li>
<li>
<p>If you do not want to use the amplicon to genome database that you downloaded in Configs or Database modules, you can point to the database directory of your interest as well by <code>--amplicon-to-genome-db</code> or <code>-a</code></p>
</li>
<li>
<p>From each sample you can choose the top k abundant taxa to be selected for downloading genome, for instance <code>--k 10</code> will select the top 10 taxa from each sample</p>
</li>
<li>
<p>When selecting genomes you need a precentage of similarity cutoff to select a representaive genome from GTDB. You do this by <code>--similarity 96</code>. This will set 96% as a similarity cutoff.</p>
</li>
<li>
<p>Finally, you need to provide the directory where you want the representative genomes to be saved. Additionally, a few extra files 
providing information about the fetched genomes will be saved here:</p>
</li>
<li>
<p><code>GenomeAccessions.csv</code> provides the NCBI accession IDs of the found genomes.</p>
</li>
<li><code>SelectedFeatures.csv</code> provides the feature IDs, hashes, from QIIME outputs for the fetched genomes.</li>
<li><code>TopKTaxa.csv</code> taxonomy name of the replicons for which a genome has been found.</li>
<li><code>Amplicon2Genome_OutInfo.json</code> <strong><em>important</em></strong> This JSON file is a metadata about the genomes that were found. This is later used to align these genomes to the protein database.</li>
</ul>
<p>A complete amplicon-to-genome command will look like:</p>
<pre><code>
ADToolbox Metagenomics amplicon-to-genome -f ~/Desktop/test/feature-table.tsv \
-r ~/Desktop/test/dna-sequences.fasta \
-t ~/Desktop/test/taxonomy.tsv \
-o ~/Desktop/test/ --k 10 \
-q ~/Desktop/test/ \
-a ~/Desktop/ADToolbox/Database/Amplicon2GenomeDBs/

</code></pre>
<hr />
<ul>
<li>align-genomes This submodule uses MMseqs to align a list of genomes to the protein database of ADToolbox for functional analysis.</li>
</ul>
<p>You can run this submodule by the following arguments:</p>
<pre><code>
ADToolbox Metagenomics align-genomes --help    
────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox Metagenomics align-genomes [-h] [-i INPUT_FILE]
                                            [-d PROTEIN_DB_DIR]
                                            [-o OUTPUT_DIR] [-b BIT_SCORE]
                                            [-e E_VALUE]

options:
  -h, --help            show this help message and exit
  -i INPUT_FILE, --input-file INPUT_FILE
                        Input the address of the JSON file includeing
                        information about the genomes to be aligned
  -d PROTEIN_DB_DIR, --protein-db-dir PROTEIN_DB_DIR
                        Directory containing the protein database to be used
                        for alignment
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        Output the directory to store the alignment results
  -b BIT_SCORE, --bit-score BIT_SCORE
                        Minimum Bit Score cutoff for alignment
  -e E_VALUE, --e-value E_VALUE
                        Minimum e-vlaue score cutoff for alignment

</code></pre>
<ul>
<li>
<p>input-file: you need to give a JSON file similar to what you create in amplicon-to-genome, see previous submodule, so that ADToolbox finds
all the information it needs about the genomes</p>
</li>
<li>
<p>protein-db-dir: you can provide ADToolboxes protein database fasta file, or any protein database that you would like to align your genomes with.</p>
</li>
<li>output-dir: Describes the location to save the alignment results. </li>
<li>bits-core: Minimum bit score to filter out the alignment results.</li>
<li>e_value: Minimum bit score to filter out the alignment results.</li>
</ul>
<p>The outputs of this step includes one more import file:</p>
<p><code>Alignment_Info.json</code>  -&gt; This file is used by map-genomes-to-adm</p>
<p>A complete command for this submodule would looklike:</p>
<pre><code>
ADToolBox Metagenomics align-genomes \
--input-file ~/Desktop/ADToolbox/Genomes/Amplicon2Genome_OutInfo.json
--protein-db-dir ~/Desktop/ADToolbox/Database/Protein_DB.fasta
--output-dir ~/Desktop/ADToolbox/Outputs/ \
--bit-score 40 \
--e-value 0.000001

</code></pre>
<hr />
<ul>
<li>make-json-from-genomes Sometimes you have genomes either from assembly or downloading it manually. In this case you can import your genomes 
to the pipeline by making a JSON file similar to  <code>Amplicon2Genome_OutInfo.json</code>. To this you need to make a CSV file for your genomes:</li>
</ul>
<pre><code>
ADToolbox Metagenomics make-json-from-genomes  --help

────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox Metagenomics make-json-from-genomes [-h] -i INPUT_FILE -o
                                                     OUTPUT_FILE

options:
  -h, --help            show this help message and exit
  -i INPUT_FILE, --input-file INPUT_FILE
                        Input the address of the CSV file includeing
                        information about the genomes to be aligned
  -o OUTPUT_FILE, --output-file OUTPUT_FILE
                        Output the directory to store the JSON file.

</code></pre>
<ul>
<li>input-file: This input file should be the address to a CSV file exactly in the following column format:</li>
</ul>
<table>
<thead>
<tr>
<th>Genome_ID</th>
<th>NCBI_Name</th>
<th>Genome_Dir</th>
</tr>
</thead>
<tbody>
<tr>
<td>Genome1</td>
<td>xyz</td>
<td>~/Desktop/...</td>
</tr>
</tbody>
</table>
<p>1- Genome_ID: Identifier for the genome, preferrably; NCBI ID
2- NCBI_Name: NCBI taxonomy name for the genome: Does not need to be in a specific format
3- Genome_Dir: Absolute path to the fasta files: NOT .gz</p>
<ul>
<li>output-file: output directory for the generated JSON to be saved. </li>
</ul>
<p>An example of this command would be:</p>
<pre><code>
ADToolBox Metagenomics make-json-from-genomes --input-file ~/Desktop/MyGenomes.CSV --output-files ~/Desktop/Genome_info.json

</code></pre>
<hr />
<ul>
<li>map-genomes-to-adm: This command will take a JSON file that has Alignment info for all of the genomes, and will output the mapping to ADM models. This is how you run this command:</li>
</ul>
<pre><code>
ADToolbox Metagenomics map-genomes-to-adm --help
────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox Metagenomics map-genomes-to-adm [-h] [-i INPUT_FILE]
                                                 [-m MODEL] [-o OUTPUT_DIR]

options:
  -h, --help            show this help message and exit
  -i INPUT_FILE, --input-file INPUT_FILE
                        Input the address of the JSON file includeing
                        information about the alignment of the genomes to the
                        protein database
  -m MODEL, --model MODEL
                        Model determines which mapping system you'd like to
                        use; Current options: 'Modified_ADM_Reactions'
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        address to store the JSON report to be loaded with a
                        model



</code></pre>
<ul>
<li>input-file: This should be the alignment info JSON file created in align-genome step.</li>
<li>model: this must be a string defining the model that you want to map the aligned reactions to. Current option is only
"Modified_ADM_Reactions"</li>
<li>output-dir: Directory to save the Genome alignment report, Metagenome report, in JSON format. This can be used later in ADM module to include the role of different genomes in the process.</li>
</ul>
<p>An example of a complete command for this submodule is:</p>
<pre><code>ADToolbox Metagenomics map-genomes-to-adm -i ~/Desktop/alignment_info.json -m Modified_ADM_Reactions -o ~/Desktop/ADM_Mapping_Report.json

</code></pre>
<hr />
<h3 id="4-adm-module">4. ADM Module</h3>
<p>ADM module provides all the tools needed to run instances of ADM Model. This include the originsl ADM, Batstone et al., and the Modified-ADM suggested by the Authors of ADToolbox. In order to find out about all the functionalities in this module, you can run:</p>
<pre><code>
ADToolbox ADM --help
────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox ADM [-h] {original-adm1,modified-adm,show-escher-map} ...

positional arguments:
  {original-adm1,modified-adm,show-escher-map}
                        Available ADM Commands:
    original-adm1       Original ADM1:
    modified-adm        Modified ADM:
    show-escher-map     makes an escher map for modified ADM

options:
  -h, --help            show this help message and exit

</code></pre>
<ul>
<li>original-adm1: If you want to run the original ADM, batstone et al, in your browser you can run this command with the required parameters in JSON format:</li>
</ul>
<pre><code>ADToolbox ADM original-adm1 --help
────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox ADM original-adm1 [-h] [--model-parameters MODEL_PARAMETERS]
                                   [--base-parameters BASE_PARAMETERS]
                                   [--initial-conditions INITIAL_CONDITIONS]
                                   [--inlet-conditions INLET_CONDITIONS]
                                   [--reactions REACTIONS] [--species SPECIES]
                                   [--metagenome-report METAGENOME_REPORT]
                                   [--report REPORT]

options:
  -h, --help            show this help message and exit
  --model-parameters MODEL_PARAMETERS
                        Model parameters for ADM 1
  --base-parameters BASE_PARAMETERS
                        Provide json file with base parameters for original
                        ADM1
  --initial-conditions INITIAL_CONDITIONS
                        Provide json file with initial conditions for original
                        ADM1
  --inlet-conditions INLET_CONDITIONS
                        Provide json file with inlet conditions for original
                        ADM1
  --reactions REACTIONS
                        Provide json file with reactions for original ADM1
  --species SPECIES     Provide json file with species for original ADM1
  --metagenome-report METAGENOME_REPORT
                        Provide json file with metagenome report for original
                        ADM1
  --report REPORT       Describe how to report the results of original ADM1.
                        Current options are: 'dash' and 'csv'


</code></pre>
<p>Every argument is optional, and their role is clear from the comments in front of them, So we just provide a full example of this command:</p>
<pre><code>ADToolbox ADM original-adm1 \
--model-parameters ~/Desktop/Model_Parameters.json \
--base-parameters ~/Desktop/Base_Parameters.json \
--initial-conditions ~/Desktop/Initial_Conditions.json \
--inlet-conditions ~/Desktop/Inlet-Conditions.json \
--reactions ~/Desktop/Reactions.json
--species  ~/Desktop/Species.json
--metagenome-report ~/Desktop/ADM_Mapping_Report.json
--repor dash

</code></pre>
<p><strong>NOTE</strong> if you choose dash for your report, the CLI will prompt you to open your browser in the instructed address, if you choose csv, it will generate a CSV file that includes concentration profiles simulated over time.</p>
<hr />
<ul>
<li>modified-adm: This command is exactly similar to the previous one, except that it requires parameters taylored for modified ADM:</li>
</ul>
<pre><code>
ADToolbox ADM modified-adm --help 
────────────────────────────────── ADToolBox ───────────────────────────────────
usage: ADToolBox ADM modified-adm [-h] [--model-parameters MODEL_PARAMETERS]
                                  [--base-parameters BASE_PARAMETERS]
                                  [--initial-conditions INITIAL_CONDITIONS]
                                  [--inlet-conditions INLET_CONDITIONS]
                                  [--reactions REACTIONS] [--species SPECIES]
                                  [--metagenome-report METAGENOME_REPORT]
                                  [--report REPORT]

options:
  -h, --help            show this help message and exit
  --model-parameters MODEL_PARAMETERS
                        Model parameters for Modified ADM
  --base-parameters BASE_PARAMETERS
                        Provide json file with base parameters for modified
                        ADM
  --initial-conditions INITIAL_CONDITIONS
                        Provide json file with initial conditions for modified
                        ADM
  --inlet-conditions INLET_CONDITIONS
                        Provide json file with inlet conditions for modified
                        ADM
  --reactions REACTIONS
                        Provide json file with reactions for modified ADM
  --species SPECIES     Provide json file with species for modified ADM
  --metagenome-report METAGENOME_REPORT
                        Provide json file with metagenome report for modified
                        ADM
  --report REPORT       Describe how to report the results of modified ADM.
                        Current options are: 'dash' and 'csv'

</code></pre>
<p>The usage is exactly the same as the original-adm</p>
<hr />
<ul>
<li>show-escher-map: This command will prompt you to open an escher map for the modified-adm model in your browser with the instructed address:</li>
</ul>
<pre><code>ADToolbox ADM show-escher-map

</code></pre>
<hr />
<h3 id="5-documentations-module">5. Documentations Module</h3>
<p>You can view the documentaion in your CLI using rich's markdown render. You can do this by:</p>
<pre><code>ADToolbox Documentations --show 

</code></pre>
<hr />
<h3 id="6-report-module">6. Report Module</h3>
<hr />
<h3 id="7-utility-module">7. Utility Module</h3>

              
            </article>
            
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../Installation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Installation" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Installation
            </div>
          </div>
        </a>
      
      
        
        <a href="../API/" class="md-footer__link md-footer__link--next" aria-label="Next: API" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              API
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.85cb4492.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.431f3d41.min.js"></script>
      
    
  </body>
</html>